kernel/src/log_reader/checkpoint_manifest.rs

```diff
@@ -0,0 +1,262 @@
+                std::slice::from_ref(&manifest.location),
+                MANIFEST_READ_SCHMEA.clone(),
+                None,
+            )?,
+            "parquet" => engine.parquet_handler().read_parquet_files(
+                std::slice::from_ref(&manifest.location),
+                MANIFEST_READ_SCHMEA.clone(),
+                None,
+            )?,
+            extension => {
+                return Err(Error::generic(format!(
+                    "Unsupported checkpoint extension: {extension}",
+                )));
+            }
+        };
+
+        let actions = Box::new(actions.map(|batch_res| Ok(ActionsBatch::new(batch_res?, false))));
+        Ok(Self {
+            actions,
+            sidecar_visitor: SidecarVisitor::default(),
+            log_root,
+            is_complete: false,
+            manifest_file: manifest.location.clone(),
+        })
+    }
+
+    /// Extract the sidecars from the manifest file if there were any.
+    /// NOTE: The iterator must be completely exhausted before calling this
+    #[allow(unused)]
+    pub(crate) fn extract_sidecars(self) -> DeltaResult<Vec<FileMeta>> {
+        require!(
+            self.is_complete,
+            Error::generic(format!(
+                "Cannot extract sidecars from in-progress ManifestReader for file: {}",
+                self.manifest_file.location
+            ))
+        );
+
+        let sidecars: Vec<_> = self
+            .sidecar_visitor
+            .sidecars
+            .into_iter()
+            .map(|s| s.to_filemeta(&self.log_root))
+            .try_collect()?;
+
+        Ok(sidecars)
+    }
+}
+
+impl Iterator for CheckpointManifestReader {
+    type Item = DeltaResult<ActionsBatch>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let result = self.actions.next().map(|batch_result| {
+            let batch = batch_result?;
+            self.sidecar_visitor.visit_rows_of(batch.actions())?;
+            Ok(batch)
+        });
+
+        if result.is_none() {
+            self.is_complete = true;
+        }
+
+        result
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::arrow::array::{Array, StringArray, StructArray};
+    use crate::engine::arrow_data::EngineDataArrowExt as _;
+    use crate::utils::test_utils::{assert_result_error_with_message, load_test_table};
+    use crate::SnapshotRef;
+
+    use itertools::Itertools;
+    use std::sync::Arc;
+
+    /// Helper function to test manifest phase with expected add paths and sidecars
+    fn verify_manifest_phase(
+        engine: Arc<dyn Engine>,
+        snapshot: SnapshotRef,
+        expected_add_paths: &[&str],
+        expected_sidecars: &[&str],
+    ) -> DeltaResult<()> {
+        let log_segment = snapshot.log_segment();
+        let log_root = log_segment.log_root.clone();
+        assert_eq!(log_segment.checkpoint_parts.len(), 1);
+        let checkpoint_file = &log_segment.checkpoint_parts[0];
+        let mut manifest_phase =
```

Comment:
nit: variable name doesn't make sense anymore :)
```suggestion
        let mut checkpoint_manifest_reader =
```